pkg/cli/alpha/config-gen/types.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configgen
+
+import (
+	"io/ioutil"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+)
+
+// KubebuilderConfigGen implements the API for generating configuration
+type KubebuilderConfigGen struct {
+	metav1.TypeMeta   `json:",inline" yaml:",omitempty"`
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Spec is the configuration spec defining what configuration should be produced.
+	Spec KubebuilderConfigGenSpec `json:"spec,omitempty" yaml:"spec,omitempty"`
+
+	// Status is the configuration status defined at runtime.
+	Status KubebuilderConfigGenStatus `json:"status,omitempty" yaml:"status,omitempty"`
+
+	// Project is for the field name used by the PROJECT file (config), instead of spec.
+	// Use Spec unless in the PROJECT file.
+	Project *KubebuilderConfigGenSpec `json:"config,omitempty" yaml:"config,omitempty"`
+}
+
+var (
+	// NamespaceComponent will create the namespace for the controller-manager
+	NamespaceComponent = "namespace"
+
+	// ControllerManagerComponent will create the controller-manager Deployment
+	ControllerManagerComponent = "controller-manager"
+
+	// WebhooksComponent will create the webhook configurations
+	WebhooksComponent = "webhooks"
+
+	// CRDsComponent will create the CRDs
+	CRDsComponent = "crds"
+
+	// RBACComponent will create RBAC rules
+	RBACComponent = "rbac"
+
+	// CertManagerComponent will create the Issuer and Certificate resources
+	// for CertManager to inject the certificates
+	CertManagerComponent = "cert-manager"
+
+	// PrometheusComponent will create the ServiceMonitor resource
+	PrometheusComponent = "prometheus"
+)
```

Comment:
This seems like a good candidate for an enum.

Reply:
Do you mean with `iota` or something else?  These are values users will specify.

Reply:
It can be a subtype of `string`, doesn't need to be `int` using `iota`. How does the user specify this? If it does through flags there is probably someway to provide a list of valid values to a flag.

Reply:
Nevermind about the enum, `const` should be better than `var` though.

Reply:
We should drop these if we are going to make them fields